prymer/api/picking.py

```diff
@@ -186,285 +79,108 @@ def score(
+    exceeding the maximum will also be discarded.
+
     Args:
         left_primers: the left primers
         right_primers: the right primers
         target: the genome mapping for the target
-        params: the parameters used for filtering
+        amplicon_sizes: minimum, optimal, and maximum amplicon sizes (lengths)
+        amplicon_tms: minimum, optimal, and maximum amplicon Tms
+        max_heterodimer_tm: if supplied, heterodimer Tms will be calculated for primer pairs,
+          and those exceeding the maximum Tm will be discarded
+        weights: the set of penalty weights
         fasta: the FASTA file from which the amplicon sequence will be retrieved.
 
     Returns:
-        the list of primer pairs, sorted by penalty (increasing)
+        an iterator over all the valid primer pairs, unsorted
     """
-    # generate all the primer pairs
-    primer_pairs: list[PrimerPair] = []
-    for left_primer in left_primers:
-        for right_primer in right_primers:
-            if left_primer.span.refname != right_primer.span.refname:
-                raise ValueError(
-                    "Cannot create a primer pair from left and right primers on different"
-                    f" references: left: '{left_primer.span.refname}'"
-                    f" right: '{right_primer.span.refname}'"
+    # Short circuit if we have no left primers or no right primers
+    if not any(left_primers) or not any(right_primers):
+        return
+
+    if not all(p.span.refname == target.refname for p in left_primers):
+        raise ValueError("Left primers exist on different reference than target.")
+
+    if not all(p.span.refname == target.refname for p in right_primers):
+        raise ValueError("Left primers exist on different reference than target.")
+
+    # Grab the sequence we'll use to fill in the amplicon sequence
+    region_start = min(p.span.start for p in left_primers)
+    region_end = max(p.span.end for p in right_primers)
+    bases = fasta.fetch(target.refname, region_start - 1, region_end)
+
+    with NtThermoAlign() as ntthal:
+        # generate all the primer pairs that don't violate hard size and Tm constraints
+        for lp in left_primers:
+            for rp in right_primers:
+                if rp.span.end - lp.span.start + 1 > amplicon_sizes.max:
+                    continue
+
+                amp_mapping = Span(refname=target.refname, start=lp.span.start, end=rp.span.end)
+                amp_bases = bases[
+                    amp_mapping.start - region_start : amp_mapping.end - region_start + 1
+                ]
+                amp_tm = calculate_long_seq_tm(amp_bases)
+
+                if amp_tm < amplicon_tms.min or amp_tm > amplicon_tms.max:
+                    continue
+
+                if max_heterodimer_tm is not None:
+                    if ntthal.duplex_tm(lp.bases, rp.bases) > max_heterodimer_tm:
+                        continue
+
+                penalty = score(
```

Comment:
**question** Naive question - why is `duplex_tm()` a method on `NtThermoAligner`, but `score()` and `calculate_long_seq_tm()` are standalone functions? Are the latter accessible through `primer3`?

Reply:
`duplex_tm()` is a complicated thing that does a thermodynamic alignment of two sequences and predicts their Tm, which is why we delegate to ntthal instead of trying to replicate that.

`score()` is mimicking something in primer3, but is not easily accessible independent of designing primer pairs in primer3.

`calculate_long_seq_tm()` is so simple it's not worth trying to redirect to primer3.